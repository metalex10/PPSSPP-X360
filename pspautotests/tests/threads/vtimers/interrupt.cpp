#include "shared.h"

SceUID vtimer;
bool intrRan = false;

extern "C" void interruptFunc(int no, void *arg) {
	SceKernelVTimerInfo info;
	info.size = sizeof(info);

	if (intrRan) {
		return;
	}
	intrRan = true;
	checkpoint("  ** Inside interrupt");

	checkpoint("sceKernelCreateVTimer: %08x", sceKernelCreateVTimer("test", NULL));
	checkpoint("sceKernelSetVTimerTimeWide: %016llx", sceKernelSetVTimerTimeWide(vtimer, 0));
	checkpoint("sceKernelGetVTimerBaseWide: %016llx", sceKernelGetVTimerBaseWide(vtimer));
	checkpoint("sceKernelGetVTimerTimeWide: %016llx", sceKernelGetVTimerTimeWide(vtimer));
	checkpoint("sceKernelStartVTimer: %08x", sceKernelStartVTimer(vtimer));
	checkpoint("sceKernelStopVTimer: %08x", sceKernelStopVTimer(vtimer));
	checkpoint("sceKernelSetVTimerHandler: %08x", sceKernelSetVTimerHandler(vtimer, NULL, NULL, NULL));
	checkpoint("sceKernelCancelVTimerHandler: %08x", sceKernelCancelVTimerHandler(vtimer));
	checkpoint("sceKernelReferVTimerStatus: %08x", sceKernelReferVTimerStatus(vtimer, &info));
	checkpoint("sceKernelDeleteVTimer 0: %08x", sceKernelDeleteVTimer(0));
	checkpoint("sceKernelDeleteVTimer: %08x", sceKernelDeleteVTimer(vtimer));
}

extern "C" int main(int argc, char *argv[]) {
	SceKernelVTimerInfo info;
	info.size = sizeof(info);

	vtimer = sceKernelCreateVTimer("test", NULL);

	checkpointNext("Inside interrupt:");
	checkpoint("sceKernelRegisterSubIntrHandler 1: %08x", sceKernelRegisterSubIntrHandler(PSP_VBLANK_INT, 1, (void *)interruptFunc, NULL));
	checkpoint("sceKernelEnableSubIntr: %08x", sceKernelEnableSubIntr(PSP_VBLANK_INT, 1));
	checkpoint("sceKernelDelayThread: %08x", sceKernelDelayThread(30000));
	checkpoint("sceKernelDisableSubIntr: %08x", sceKernelDisableSubIntr(PSP_VBLANK_INT, 1));
	checkpoint("sceKernelReleaseSubIntrHandler: %08x", sceKernelReleaseSubIntrHandler(PSP_VBLANK_INT, 1));

	sceKernelSetVTimerTimeWide(vtimer, 0);
	checkpointNext("Interrupts disabled:");
	int flag = sceKernelCpuSuspendIntr();
	SceUID vtimer2 = sceKernelCreateVTimer("test", NULL);
	checkpoint("sceKernelCreateVTimer: %08x", vtimer2 >= 0 ? 0x1337 : vtimer2);
	checkpoint("sceKernelSetVTimerTimeWide: %016llx", sceKernelSetVTimerTimeWide(vtimer, 0));
	checkpoint("sceKernelGetVTimerBaseWide: %016llx", sceKernelGetVTimerBaseWide(vtimer));
	checkpoint("sceKernelGetVTimerTimeWide: %016llx", sceKernelGetVTimerTimeWide(vtimer));
	checkpoint("sceKernelStartVTimer: %08x", sceKernelStartVTimer(vtimer));
	checkpoint("sceKernelStopVTimer: %08x", sceKernelStopVTimer(vtimer));
	checkpoint("sceKernelSetVTimerHandler: %08x", sceKernelSetVTimerHandler(vtimer, NULL, NULL, NULL));
	checkpoint("sceKernelCancelVTimerHandler: %08x", sceKernelCancelVTimerHandler(vtimer));
	checkpoint("sceKernelReferVTimerStatus: %08x", sceKernelReferVTimerStatus(vtimer, &info));
	checkpoint("sceKernelDeleteVTimer 0: %08x", sceKernelDeleteVTimer(0));
	checkpoint("sceKernelDeleteVTimer: %08x", sceKernelDeleteVTimer(vtimer));
	sceKernelCpuResumeIntr(flag);

	sceKernelDeleteVTimer(vtimer);
	return 0;
}