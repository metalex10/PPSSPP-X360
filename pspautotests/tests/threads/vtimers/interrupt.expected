[x] Inside interrupt:
[x] sceKernelRegisterSubIntrHandler 1: 00000000
[x] sceKernelEnableSubIntr: 00000000
[r]   ** Inside interrupt
[x] sceKernelCreateVTimer: 80020064
[x] sceKernelSetVTimerTimeWide: ffffffffffffffff
[x] sceKernelGetVTimerBaseWide: 0000000000000000
[x] sceKernelGetVTimerTimeWide: 0000000000000000
[x] sceKernelStartVTimer: 00000000
[x] sceKernelStopVTimer: 00000001
[x] sceKernelSetVTimerHandler: 00000000
[x] sceKernelCancelVTimerHandler: 00000000
[x] sceKernelReferVTimerStatus: 00000000
[x] sceKernelDeleteVTimer 0: 80020064
[x] sceKernelDeleteVTimer: 80020064
[x] sceKernelDelayThread: 00000000
[x] sceKernelDisableSubIntr: 00000000
[x] sceKernelReleaseSubIntrHandler: 00000000

[x] Interrupts disabled:
[x] sceKernelCreateVTimer: 00001337
[x] sceKernelSetVTimerTimeWide: 0000000000000000
[x] sceKernelGetVTimerBaseWide: 0000000000000000
[x] sceKernelGetVTimerTimeWide: 0000000000000000
[x] sceKernelStartVTimer: 00000000
[x] sceKernelStopVTimer: 00000001
[x] sceKernelSetVTimerHandler: 00000000
[x] sceKernelCancelVTimerHandler: 00000000
[x] sceKernelReferVTimerStatus: 00000000
[x] sceKernelDeleteVTimer 0: 800201be
[x] sceKernelDeleteVTimer: 00000000
